### ISSUE TITLE
How to calculate p-values for my coefficients?

### ISSUE THREAD

**Issue by JohannesWiesner** [resample-project/resample#113]
I am not sure if this question might be better placed on StackOverflow, but I guess it could be also a documentation issue so I guess it could also fit here. Besides, the code below might also help to explain what I meant with issue #112, so I hope, I can post it here on GitHub. I fitted a multinomial logistic regression to my data. Now I would like to get 95%-CI intervals for my beta-coefficients and additional p-values for each of them (and for each class, since with multinomial LR you get coefficients for each classes). How would I do that with your package?

Here's some example code (as written above, besides from this issue, it might also help to explain what I described in #112):

```python
import numpy as np
from sklearn.datasets import load_iris

from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline

from resample.bootstrap import bootstrap

# set random number generator
rng = np.random.RandomState(42)

# load data
X,y = load_iris(return_X_y=True)

# implement Pipeline
# center data and fit multinomial logistic regression
pipe = Pipeline([('scaler',StandardScaler()),
                  ('lr',LogisticRegression())])


pipe.fit(X,y)

# prepare data for bootstrapping with resample package
# NOTE: This is a little tedious because we have to concatenate X an y first
# to one array only to separate them again a few steps later within the function
A = np.concatenate((y.reshape(-1,1),X),axis=1)

# NOTE: Currently, the resample package does not offer to provide any further
# arguments that are passed to the function besides A. So we have to rely on 
# global scope (which might be a little bit unsafe?)
def fit_mlr(A):
    
    X = A[:,1:]
    y = A[:,0]
    pipe.fit(X,y)

    return {'coef': pipe._final_estimator.coef_,
            'intercept': pipe._final_estimator.coef_}

boot_coef = bootstrap(fit_mlr,sample=A,size=10)


```

**Comment by dsaxton** [comment]
@JohannesWiesner theoretically one way to calculate p-values could be to simulate a null distribution for each coefficient by shuffling X and y with respect to each other and fitting your model many times. In this scenario you _know_ there's no relationship between X and y and you can look at the coefficient distributions in this case and compare them to what you got with the original unshuffled data to get a p-value (your p-value would be how often in your simulations you got coefficients more extreme than the actual observed one).

However, be careful when doing this because it's easy to fall into a trap of testing a hypothesis using the same that generated that hypothesis (if you've done model selection prior to testing your coefficients for example), which is a fallacy: https://en.wikipedia.org/wiki/Testing_hypotheses_suggested_by_the_data

**Comment by HDembinski** [comment]
Using global state is not unsafe in a small skript. In a large skript, you can use a functor object instead, which uses class state instead of global state.

**Comment by HDembinski** [comment]
@dsaxton thanks for this cool reference to the relevant Wikipedia article

**Comment by JohannesWiesner** [comment]
@dsaxton Ah interesting! So if I got you right, you wouldn't even recommend using `resample.bootstrap.bootstrap` to estimate confidence intervals and/or p-values for the beta-coefficients but rather some sort of randomization test? The more I get into this topic the more I see that the nomenclature seems to be a little unclear when people are talking about resampling procedures in general, but from what you said it seems that you were referring to what is described in this [Wikipedia Article](https://en.wikipedia.org/wiki/Permutation_test)?

> However, be careful when doing this because it's easy to fall into a trap of testing a hypothesis using the same data that generated that hypothesis (if you've done model selection prior to testing your coefficients for example), which is a fallacy: https://en.wikipedia.org/wiki/Testing_hypotheses_suggested_by_the_data

Thanks for the link! Yes, I kept that in my mind by splitting my dataset into a train and test set.

**Comment by dsaxton** [comment]
> So if I got you right, you wouldn't even recommend using `resample.bootstrap.bootstrap` to estimate confidence intervals and/or p-values for the beta-coefficients but rather some sort of randomization test?

It's fine to use for calculating p-values (that's one of the applications), I was just making a general comment that these tests can be made invalid pretty easily if they're part of a model-building process

**Comment by HDembinski** [comment]
Thinking more about it, I don't know what kind of p-value you want to compute. A p-value is a deviation from the null hypothesis. How you compute it depends on what the null hypothesis in your case is. I am closing this as unclear, but feel free to reopen if needed.
